src/effects/effectchain.cpp

```diff
@@ -151,37 +155,63 @@ void EffectChain::setEnabled(bool enabled) {
 }
 
 void EffectChain::enableForChannel(const ChannelHandleAndGroup& handle_group) {
-    if (!m_bAddedToEngine) {
+    // TODO(Be): remove m_enabledChannels from this class and move this logic
+    // to EffectChainSlot
+    bool bWasAlreadyEnabled = m_enabledChannels.contains(handle_group);
+    if (!bWasAlreadyEnabled) {
+        m_enabledChannels.insert(handle_group);
+    }
+
+    // The allocation of EffectStates below may be expensive, so avoid it if
+    // not needed.
+    if (!m_bAddedToEngine || bWasAlreadyEnabled) {
         return;
     }
-    if (!m_enabledChannels.contains(handle_group)) {
-        m_enabledChannels.insert(handle_group);
 
-        EffectsRequest* request = new EffectsRequest();
-        request->type = EffectsRequest::ENABLE_EFFECT_CHAIN_FOR_CHANNEL;
-        request->pTargetChain = m_pEngineEffectChain;
-        request->channel = handle_group.handle();
-        m_pEffectsManager->writeRequest(request);
+    EffectsRequest* request = new EffectsRequest();
+    request->type = EffectsRequest::ENABLE_EFFECT_CHAIN_FOR_INPUT_CHANNEL;
+    request->pTargetChain = m_pEngineEffectChain;
+    request->channel = handle_group.handle();
+
+    // Allocate EffectStates here in the main thread to avoid allocating
+    // memory in the realtime audio callback thread.
+    auto pStatesForEffectsInChain = new QList<ChannelHandleMap<EffectState*>*>;
```

Comment:
This is leaking. 
Is it really necessary to have a list here, or can we pass the EffectStates one by one?

Reply:
Yes, it is necessary to have a list. It is a list of maps, where each map is for a different effect in the chain. I will add a comment clarifying that.

Reply:
I was wondering about deleting this. Would std::unique_ptr be appropriate?

Reply:
It will prevent to copy the message. You can give it a try.
Alternative each input hash can get a next pointer, to chain all required states together.
The message only needs plain pointer to the first.

Reply:
Would using std::array instead of QList help?